src/wagtail_vector_index/index/models.py

```diff
@@ -0,0 +1,402 @@
+                self.get_converter().generate_embeddings(
+                    instance, embedding_backend=self.embedding_backend
+                )
+
+        return super().rebuild_index()
+
+    def get_documents(self) -> Iterable[Document]:
+        querysets = self._get_querysets()
+
+        for queryset in querysets:
+            instances = queryset.prefetch_related("embeddings")
+            yield from self.get_converter().bulk_to_documents(
+                instances, embedding_backend=self.embedding_backend
+            )
+
+
+class PageVectorIndex(ModelVectorIndex):
+    """A model vector indexed for use with Wagtail pages that automatically
+    restricts indexed models to live pages."""
+
+    querysets: Sequence[PageQuerySet]
+
+    def _get_querysets(self) -> list[PageQuerySet]:
+        qs_list = super()._get_querysets()
+
+        # Technically a manager instance, not a queryset, but we want to use the custom
+        # methods.
+        return [cast(PageQuerySet, qs).live() for qs in qs_list]
+
+
+# ###########
+# Classes related to automatic generation of indexes for models
+# ###########
+
+
+class GeneratedIndexMixin(models.Model):
+    """Mixin for Django models that automatically generates registers a VectorIndex for the model"""
+
+    vector_index_class: ClassVar[Optional[type[VectorIndex]]] = None
+    converter_class: ClassVar[type[DocumentConverter]]
+
+    class Meta:
+        abstract = True
+
+    @classmethod
+    def get_vector_index(cls):
+        """Get a vector index instance for this model"""
+
+        # If the user has specified a custom `vector_index_class`, use that
+        if cls.vector_index_class:
+            index_cls = cls.vector_index_class
+        # If the model is a Wagtail Page, use a special PageVectorIndex
+        elif issubclass(cls, Page):
+            index_cls = PageVectorIndex
+        # Otherwise use the standard ModelVectorIndex
+        else:
+            index_cls = ModelVectorIndex
+
+        return type(
+            f"{cls.__name__}Index",
+            (index_cls,),
+            {
+                "querysets": [cls.objects.all()],
+                "converter_class": cls.converter_class,
+            },
+        )()
+
+
+class VectorIndexedMixin(EmbeddableFieldsMixin, GeneratedIndexMixin, models.Model):
+    """Model mixin which adds both the embeddable fields behaviour and the automatic index behaviour to a model."""
+
+    converter_class: ClassVar[
+        type[EmbeddableFieldsDocumentConverter]
+    ] = EmbeddableFieldsDocumentConverter
+
+    class Meta:
+        abstract = True
+
+
+def register_indexed_models():
+    """Discover and register all models that are a subclass of GeneratedIndexMixin."""
+    indexed_models = [
+        model
+        for model in apps.get_models()
+        if issubclass(model, GeneratedIndexMixin) and not model._meta.abstract
+    ]
+    for model in indexed_models:
+        registry.register()(model.get_vector_index().__class__)
```

Comment:
I think we should say `instantiate_vector_index`, where `get` operation should never mutate or create anything.

Why do we instantiate vector index just to get a class?

Should this be `model.get_vector_index_class()`?

Reply:
Above refactoring has resolved this.